src/auth/mod.ts

```diff
@@ -13,35 +15,169 @@ interface DecodedToken {
   };
 }
 
+interface JWK {
+  kid: string;
+  use: string;
+  kty: string;
+  alg: string;
+  n: string;
+  e: string;
+}
+
 const logger = getLogger(import.meta);
 
-function isSubInPodsConfig(payload: DecodedToken): boolean {
-  return podsConfig.pods.some((pod) =>
-    pod.sub === payload.sub &&
-    pod.serviceaccount.uid === payload["kubernetes.io"]?.serviceaccount.uid
-  );
+export async function verifyServiceAccountToken(
+  token: string,
+): Promise<string> {
+  const payload = await verifyToken(token);
+
+  if (
+    !payload["kubernetes.io"] || !payload["kubernetes.io"].serviceaccount.name
+  ) {
+    const message = "Payload does not contain service account name field";
+    logger.error(message);
+    throw new HTTPException(HTTP_STATUS_CODES.UNAUTHORIZED, {
+      message,
+    });
+  }
+
+  return payload["kubernetes.io"].serviceaccount.name;
+}
+
+let jwks: JWK[] | undefined;
+const jwkExpiration = 24 * 60 * 60 * 1000; // 24 hours
+let expirationTime = Date.now() + jwkExpiration;
+
+function isJWKExpired(): boolean {
+  return Date.now() > expirationTime;
 }
 
-export function decodeToken(token: string): DecodedToken | null {
-  if (token.startsWith("Bearer ")) {
-    token = token.slice(7);
+const cryptoKeys: Map<string, CryptoKey> = new Map();
+async function verifyToken(token: string): Promise<DecodedToken> {
+  if (!jwks || isJWKExpired()) {
+    jwks = await getKeys();
   }
-  const [_headers, payload, _signature] = decode(token);
 
-  const data = payload as DecodedToken;
-  if (!data || typeof data !== "object" || !data.sub) {
-    logger.warn("Invalid token");
-    throw new HTTPException(401, {
-      message: "Invalid token",
-    });
+  const header = decode(token)[0];
+  const kid = (header as { kid?: string })?.kid;
+  if (!kid) {
+    const message = "Missing kid in token header";
+    logger.error(message);
+    throw new HTTPException(401, { message });
+  }
+  const matchingKey = jwks.find((k) => k.kid === kid);
+  if (!matchingKey) {
+    const message = "Key not found for kid in JWKs";
+    logger.error(message);
+    throw new HTTPException(401, { message });
+  }
+  const key = JSON.stringify(matchingKey);
+
+  const cryptoKey = cryptoKeys.has(kid)
+    ? cryptoKeys.get(kid)!
+    : await crypto.subtle.importKey(
```

Comment:
can we cache the loaded key? I would split the function in 2 : 1 that does take care of the caching of the keys and then have a object (for direct access o(1), and avoid looping o(n)) to get the matching key. As you load the key on the start, there is no need for something null, rather the object could be empty until the key is available and you can block until you have the first ones loaded